### PR TITLE
Detect divergence and report it with the convergence status in the metrics file

### PR DESCRIPTION
**Pull request by artemsok**
Sometimes training diverges with average training and validation perplexities sky-rocketing. To detect this early, we can test if the average validation perplexity surpasses double the the target vocabulary size and stop training. Additionally, we report the convergence and divergence status in the metrics file to provide some more information on why training stopped: because it actually converged, diverged or stopped for some the reason, like, hitting one of the max_{epochs, updates} limits.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/training.py [hunk 1]
```diff
@@ -586,7 +588,27 @@ def fit(self,
...
+                        self.state.converged = False
+
+                    if min_updates is not None and self.state.updates < min_updates:
+                        logger.info("Minimum number of updates (%d) not reached yet: %d",
+                                    min_updates, self.state.updates)
+                        self.state.converged = False
+
+                    if min_samples is not None and self.state.samples < min_samples:
+                        logger.info("Minimum number of samples (%d) not reached yet: %d",
+                                    min_samples, self.state.samples)
```
Comment (fhieber) [inline sockeye/training.py]: is there actually a `self.state.converged = False` missing?

Comment (artemsok) [reply sockeye/training.py]: Strange. looks like it's was missing all along, in the previous code. Adding in the next commit.

File: sockeye/utils.py [hunk 2]
```diff
@@ -791,7 +791,7 @@ def write_metrics_file(metrics: List[Dict[str, Any]], path: str):
     """
     with open(path, 'w') as metrics_out:
         for checkpoint, metric_dict in enumerate(metrics, 1):
-            metrics_str = "\t".join(["{}={}".format(name, value) for name, value in sorted(metric_dict.items())])
+            metrics_str = "\t".join(["{}={:f}".format(name, value) for name, value in sorted(metric_dict.items())])
```
Comment (tdomhan) [inline sockeye/utils.py]: Why the change to fixed point?

File: sockeye/training.py [hunk 3]
```diff
@@ -747,11 +757,16 @@ def _update_metrics(self,
...
         gpu_memory_usage = utils.get_gpu_memory_usage(self.model.context)
         checkpoint_metrics['used-gpu-memory'] = sum(v[0] for v in gpu_memory_usage.values())
+        checkpoint_metrics['converged'] = self.state.converged
+        checkpoint_metrics['diverged'] = self.state.diverged
 
+        # copy values from checkpoint_metrics, skipping NaNs and infinities that break numpy's histograms
         for name, value in metric_train.get_name_value():
-            checkpoint_metrics["%s-train" % name] = value
+            if np.isfinite(value):
+                checkpoint_metrics["%s-train" % name] = value
```
Comment (tdomhan) [inline sockeye/training.py]: how about adding a warning here and below?

File: sockeye/utils.py (hunk 2 above)
Comment (artemsok) [reply sockeye/utils.py]: This is for the read_metrics() @[L780](https://github.com/awslabs/sockeye/blob/09068cdeba9be538fccd78496a6b60a26d2a995f/sockeye/utils.py#L780) from utils.py that assumes everything is float.

File: sockeye/training.py (hunk 3 above)
Comment (artemsok) [reply sockeye/training.py]: Added

File: sockeye/training.py [hunk 4]
```diff
@@ -747,11 +757,16 @@ def _update_metrics(self,
...
         for name, value in metric_val.get_name_value():
-            checkpoint_metrics["%s-val" % name] = value
+            if np.isfinite(value):
```
Comment (tdomhan) [inline sockeye/training.py]: actually what are we guarding against here? I tried invoking an issue by setting value to `np.inf`, but that didn't cause any issues. Was this related to Tensorboard? If so I think we should remove values only for tensorboard logging (or for whichever part can not handle `inf`). For the metric file itself I would argue that having the `inf` is actually valuable for debugging, no?

Comment (tdomhan) [reply sockeye/training.py]: So this seems to be caused by `add_histogram` in `TensorboardLogger`. So to guard against this problem we should add the check (+ warning) there. The check for finite for scalars we can then keep.

File: sockeye/training.py [hunk 5]
```diff
@@ -611,6 +634,13 @@ def fit(self,
                     logger.info("Validation-%s has not improved for %d checkpoints, best so far: %f",
                                 early_stopping_metric, self.state.num_not_improved, self.state.best_metric)
 
+                # (5) detect divergence with respect to perplexity value at the last checkpoint
+                if self.state.metrics and not has_improved:
+                    # perplexity cannot exceed the number of target words, using the double to avoid precision issues
+                    last_ppl_value = self.state.metrics[-1]["%s-val" % C.PERPLEXITY]
+                    if not np.isfinite(last_ppl_value) or last_ppl_value > 2 * len(self.target_vocab):
+                        self.state.diverged = True
```
Comment (tdomhan) [inline sockeye/training.py]: I think when we detect that the model diverged a logging message would be appropriate.

File: sockeye/training.py (hunk 4 above)
Comment (artemsok) [reply sockeye/training.py]: Moving the check to TensorboardLogger.log_metrics() as per offline discussion.

File: sockeye/utils.py (hunk 2 above)
Comment (artemsok) [reply sockeye/utils.py]: Now doing the check in the read_metrics().

File: sockeye/training.py (hunk 5 above)
Comment (artemsok) [reply sockeye/training.py]: Added

File: sockeye/training.py [hunk 6]
```diff
@@ -586,7 +588,28 @@ def fit(self,
                 for name, val in metric_val.get_name_value():
                     logger.info('Checkpoint [%d]\tValidation-%s=%f', self.state.checkpoint, name, val)
 
-                # (3) update training metrics
+                # (2) determine stopping
```
Comment (tdomhan) [inline sockeye/training.py]: Why did this move before the actual update of `num_not_improved`? Wouldn't lead this to models training for one checkpoint longer than before?

File: sockeye/training.py [hunk 7]
```diff
@@ -611,6 +634,13 @@ def fit(self,
                     logger.info("Validation-%s has not improved for %d checkpoints, best so far: %f",
                                 early_stopping_metric, self.state.num_not_improved, self.state.best_metric)
 
+                # (5) detect divergence with respect to perplexity value at the last checkpoint
+                if self.state.metrics and not has_improved:
+                    # perplexity cannot exceed the number of target words, using the double to avoid precision issues
```
Comment (tdomhan) [inline sockeye/training.py]: maybe phrase as "should not"

**Comment by tdomhan** [comment]
mypy seems unhappy now
```
mypy --ignore-missing-imports --follow-imports=silent @typechecked-files --no-strict-optional
sockeye/utils.py:783: error: Incompatible types in assignment (expression has type "float", target has type "bool")
The command "mypy --ignore-missing-imports --follow-imports=silent @typechecked-files --no-strict-optional" exited with 1.

```

File: sockeye/training.py (hunk 6 above)
Comment (artemsok) [reply sockeye/training.py]: Reverted to (almost) the original step order and now manually adding validation metric to the loop that sets has_improved, since it wasn't yet available from self.state.metrics

File: sockeye/training.py [hunk 8]
```diff
@@ -586,14 +588,12 @@ def fit(self,
...
-                # (3) update training metrics
-                self._update_metrics(metric_train, metric_val, process_manager)
-                metric_train.reset()
-
-                # (4) determine improvement
+                # (2) determine improvement
                 has_improved = False
                 previous_best = self.state.best_metric
-                for checkpoint, metric_dict in enumerate(self.state.metrics, 1):
+                # at this point state.self..metrics doesn't have validation results yet
```
Comment (tdomhan) [inline sockeye/training.py]: nit: `self.state.metrics`

**Comment by tdomhan** [comment]
Now there are a couple of pylint errors which seem to be unrelated (this could be a new version of pylint). Maybe we should fix pylint to the old version for Travis for now.
```
34.00s$ pylint --rcfile=pylintrc sockeye -E
************* Module sockeye.image_captioning.utils
sockeye/image_captioning/utils.py:194:38: E1133: Non-iterable value diff_shape is used in an iterating context (not-an-iterable)
The command "pylint --rcfile=pylintrc sockeye -E" exited with 2.
21.64s$ pylint --rcfile=pylintrc test -E
************* Module test.unit.test_loss
test/unit/test_loss.py:124:27: E1130: bad operand type for unary -: any (invalid-unary-operand-type)
test/unit/test_loss.py:131:27: E1130: bad operand type for unary -: any (invalid-unary-operand-type)
test/unit/test_loss.py:149:29: E1130: bad operand type for unary -: any (invalid-unary-operand-type)
************* Module test.unit.test_encoder
test/unit/test_encoder.py:202:11: E1101: Function 'any' has no 'all' member (no-member)
The command "pylint --rcfile=pylintrc test -E" exited with 2.
```

File: sockeye/training.py [hunk 9]
```diff
@@ -1061,7 +1077,10 @@ def log_metrics(self, metrics: Dict[str, Union[float, int, mx.nd.NDArray]], chec
 
         for name, value in metrics.items():
             if isinstance(value, mx.nd.NDArray):
-                self.sw.add_histogram(tag=name, values=value, bins=100, global_step=checkpoint)
+                if mx.ndarray.contrib.isfinite(value).sum() == value.size():
```
Comment (tdomhan) [inline sockeye/training.py]: when was this added to mxnet? It does not seem to be present in my version (`1.3.1`).

Comment (tdomhan) [reply sockeye/training.py]: It's mentioned in the 1.4.0 changelog: https://github.com/apache/incubator-mxnet/blob/master/NEWS.md So maybe we  should use numpy for now and add a TODO to use the mxnet version with the next release.

Comment (tdomhan) [reply sockeye/training.py]: Or simply add the three line function to `utils.py` for now https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/ndarray/contrib.py#L491

File: sockeye/training.py [hunk 10]
```diff
@@ -586,14 +588,12 @@ def fit(self,
...
-                self._update_metrics(metric_train, metric_val, process_manager)
-                metric_train.reset()
-
-                # (4) determine improvement
+                # (2) determine improvement
                 has_improved = False
                 previous_best = self.state.best_metric
-                for checkpoint, metric_dict in enumerate(self.state.metrics, 1):
+                # at this point state.self.metrics doesn't have validation results yet
+                current_checkpoint_val_metric = {"%s-val" % name:val for name, val in metric_val.get_name_value()}
```
Comment (tdomhan) [inline sockeye/training.py]: this will not actually work if the optimized metric is `BLEU` as it will only later be added to the metrics.

Comment (artemsok) [reply sockeye/training.py]: I now moved the waiting for collect_result() out of the _update_metrics() to let state.metrics have the bleu-val if it's computed. Tested locally.
